### PR TITLE
Set interface name to the network_interface name for macvlan and ipvlan networks

### PR DESCRIPTION
**Pull request by vikas-goel**
When interface_name attribute in containers.conf file is set to "device", then set interface names inside containers same as the network_interface names of the respective network.

The change applies to macvlan and ipvlan networks only. The interface_name attribute value has no impact on any other types of networks.

If the interface name is set in the user request, then that takes precedence.

The interface_name attribute in the containers.conf file was added via https://github.com/containers/common/pull/1814.

Fixes: #21313

#### Does this PR introduce a user-facing change? Yes

```release-note
A new global attribute interface_name has been added to the containers.conf file. The default value of the attribute is "". The attribute value can be overridden to "device". This setting is effective for macvlan and ipvlan networks only. When set to "device", the network interface names inside containers will be set to the network_interface property value as seen in the podman inspect of the respective network, by default. If the network_interface property is not set or the global attribute is left unmodified, then the network interface name is set to the ethX pattern as usual, by default. 
```

### CODE REVIEW

**Comment by vikas-goel** [comment]
@Luap99 , @TomSweeneyRedHat ,

The two `machine` related test case failures seem to be altogether in an independent area for unrelated reasons. For example, similar issue was seen in https://github.com/containers/podman/pull/21339/checks?check_run_id=21024221296.

What is the next step? Can we get it merged?

**Comment by vikas-goel** [comment]
Pinging @TomSweeneyRedHat for review

**Comment by vikas-goel** [comment]
Pinging @rhatdan for review

File: libpod/networking_common.go
Comment (mheon) [inline libpod/networking_common.go:487]: Again, should only be allowed for ipvlan/macvlan

**Comment by mheon** [comment]
@edsantiago Mind taking a look at the tests?

**Comment by TomSweeneyRedHat** [comment]
@Luap99 is the SME here, and I'd like to get his eyeballs on this.
LGTM otherwise.

**Comment by baude** [comment]
general comments: 

* can you comment your tests more verbosely for the future
* where is this documented
* i understand this is paired with a containers-common pr, can you please cite it
* can you please change the commit title to include something like macvlan only, in the theme of matt's feedback
* please include macvlan only in the pr description as well

well done getting through CI!  And appreciate the tests.

**Comment by vikas-goel** [comment]
> general comments:
> 
> * can you comment your tests more verbosely for the future
> * where is this documented
> * i understand this is paired with a containers-common pr, can you please cite it
> * can you please change the commit title to include something like macvlan only, in the theme of matt's feedback
> * please include macvlan only in the pr description as well
> 
> well done getting through CI! And appreciate the tests.

Incorporated all the comments. The attribute definition is documented in the c/common repo. The release notes in this PR describes the change.

File: libpod/networking_common.go
Comment (mheon) [inline libpod/networking_common.go:644]: This isn't just used for macvlan and ipvlan, though? It's used for all drivers.

**Comment by edsantiago** [comment]
What bothers me a little about the tests is that, when run against `main`,  all of them pass (or are skipped) rootless, and only one fails when run as root. If this is intentional -- like, good-citizen adding a whole bunch of tests unrelated to this PR -- yay and thank you! But if more than one of these tests are supposed to test the PR, there's something wrong.

I'll take a closer look tomorrow.

File: libpod/networking_common.go
Comment (vikas-goel) [reply libpod/networking_common.go:644]: The comment about the drivers is for the latter part. That is, network_interface name. Do you want me to change that?

**Comment by vikas-goel** [comment]
> What bothers me a little about the tests is that, when run against `main`, all of them pass (or are skipped) rootless, and only one fails when run as root. If this is intentional -- like, good-citizen adding a whole bunch of tests unrelated to this PR -- yay and thank you! But if more than one of these tests are supposed to test the PR, there's something wrong.
> 
> I'll take a closer look tomorrow.

There are 4/10 that skip in rootless, not all. They still make sense for root mode. If there is a better way, let me know.

File: libpod/networking_common.go
Comment (vikas-goel) [reply libpod/networking_common.go]: Updated.

**Comment by edsantiago** [comment]
What I'm trying to say is that most of these new tests _do not test your PR_. Are they supposed to?

**Comment by vikas-goel** [comment]
> What I'm trying to say is that most of these new tests _do not test your PR_. Are they supposed to?

Not sure I am understanding. The PR CI result shows all of them are getting tested in root mode.

**Comment by edsantiago** [comment]
1. I write code.
2. I write a test for this code.
3. **This new test must FAIL when run against the previous commit!**

That is one of the most important and fundamental elements of testing. 

Most of your new tests pass when run against main. this could be because
a. you are a kind person and have written a bunch of tests for functionality that your PR does not address, that has nothing whatsoever to do with your PR, but you found a hole in our testing and have corrected it. Or
b. your tests are not testing what you think you're testing

I do not have time or energy tonight to figure out which is the reason. So my question to you is, did you deliberately write a bunch of new tests that have nothing to do with your PR, simply to improve our coverage? Or were you expecting all of your new tests to be tests of your PR?

Thank you either way. We are all grateful for new code and new test coverage.

**Comment by vikas-goel** [comment]
> 1. I write code.
> 2. I write a test for this code.
> 3. **This new test must FAIL when run against the previous commit!**
> 
> That is one of the most important and fundamental elements of testing.
> 
> Most of your new tests pass when run against main. this could be because a. you are a kind person and have written a bunch of tests for functionality that your PR does not address, that has nothing whatsoever to do with your PR, but you found a hole in our testing and have corrected it. Or b. your tests are not testing what you think you're testing
> 
> I do not have time or energy tonight to figure out which is the reason. So my question to you is, did you deliberately write a bunch of new tests that have nothing to do with your PR, simply to improve our coverage? Or were you expecting all of your new tests to be tests of your PR?
> 
> Thank you either way. We are all grateful for new code and new test coverage.

I see what you are point.

Most of the new test cases are related existing behavior. That is, checking the ethX interface naming inside container. That is why they are passing on the previous commit. So, what you said "written a bunch of tests for functionality that your PR does not address, that has nothing whatsoever to do with your PR, but you found a hole in our testing and have corrected it." is probably correct.

File: test/e2e/container_iface_name_test.go
Comment (edsantiago) [inline test/e2e/container_iface_name_test.go:465]: Something like
```diff
                Expect(exec2).Should(ExitCleanly())
+               Expect(exec2.OutputToString()).Should(ContainSubstring("10.25.40."))
+               Expect(exec2.OutputToString()).ShouldNot(ContainSubstring("10.10.0."))
```

File: test/e2e/container_iface_name_test.go
Comment (edsantiago) [inline test/e2e/container_iface_name_test.go:108]: All these tests use the same subnets, `10.10.0` and `10.25.40`. All these tests also run in parallel. @containers/podman-maintainers will this cause collisions and flakes if two of these tests happen to run the same network commands at the same time?

**Review by edsantiago — COMMENTED** [review]
Feedback from a quick first pass. I need to switch context now, sorry.

File: test/e2e/container_iface_name_test.go
Comment (vikas-goel) [reply test/e2e/container_iface_name_test.go:85]: The commands prints a random id on stdout. So, we were to check stdout, it would be something like `!Equal("")`. Is that what you are asking for?

File: test/e2e/container_iface_name_test.go
Comment (vikas-goel) [reply test/e2e/container_iface_name_test.go:60]: Sure.

File: test/e2e/container_iface_name_test.go
Comment (vikas-goel) [reply test/e2e/container_iface_name_test.go:465]: Sure.

File: test/e2e/container_iface_name_test.go
Comment (vikas-goel) [reply test/e2e/container_iface_name_test.go:108]: That shouldn't be an issue.

File: test/e2e/container_iface_name_test.go
Comment (vikas-goel) [reply test/e2e/container_iface_name_test.go:86]: @edsantiago , your thoughts?

**Comment by vikas-goel** [comment]
@edsantiago , @mheon , @Luap99, @TomSweeneyRedHat , @rhatdan ,
How do we take it forward?

**Comment by mheon** [comment]
Once @Luap99 gets in next week and reviews, we can get through final review comments and merge.

File: libpod/networking_common.go
Comment (Luap99) [inline libpod/networking_common.go:647]: No need for the second param, just use `r.config.Containers.InterfaceName` directly here instead of forcing the caller do the right thing.

File: libpod/networking_common.go
Comment (Luap99) [inline libpod/networking_common.go:654]: Please add a comment on where "device" comes from, this should properly be a constant somewhere.

Also for macvlan and ipvlan it should use `types.MacVLANNetworkDriver` and `types.IPVLANNetworkDriver`

File: test/e2e/container_iface_name_test.go
Comment (Luap99) [inline test/e2e/container_iface_name_test.go:14]: this is not checking the exit code so it can fail silently, it needs
`Expect(session).To(Exit(0))`

File: test/e2e/container_iface_name_test.go
Comment (Luap99) [inline test/e2e/container_iface_name_test.go:19]: same here

File: test/e2e/container_iface_name_test.go
Comment (Luap99) [inline test/e2e/container_iface_name_test.go:29]: should use CONTAINERS_CONF_OVERRIDE instead

File: test/e2e/container_iface_name_test.go
Comment (Luap99) [inline test/e2e/container_iface_name_test.go:59]: This is not needed e2e tests will automatically cleanup up containers.

File: test/e2e/container_iface_name_test.go
Comment (Luap99) [inline test/e2e/container_iface_name_test.go:96]: not needed as well

File: test/e2e/container_iface_name_test.go
Comment (Luap99) [inline test/e2e/container_iface_name_test.go:114]: not needed

File: test/e2e/container_iface_name_test.go
Comment (Luap99) [inline test/e2e/container_iface_name_test.go:142]: not needed

File: test/e2e/container_iface_name_test.go
Comment (Luap99) [inline test/e2e/container_iface_name_test.go:211]: Why, this must work with netavark!

File: test/e2e/container_iface_name_test.go
Comment (Luap99) [inline test/e2e/container_iface_name_test.go:259]: same here

**Review by Luap99 — CHANGES_REQUESTED** [review]
The tests seem to carry a ton of duplication which it makes it very hard to follow I suggest rewriting them in some form of loop to deduplicate most logic.

File: libpod/networking_common.go
Comment (vikas-goel) [reply libpod/networking_common.go:654]: Will update. The "device" string is used just in this function once, nowhere else. Do we still want it defined as a constant?

**Comment by vikas-goel** [comment]
> The tests seem to carry a ton of duplication which it makes it very hard to follow I suggest rewriting them in some form of loop to deduplicate most logic.

Updated.

File: libpod/networking_common.go
Comment (vikas-goel) [reply libpod/networking_common.go]: Resolving this comment for now. If you think defining a constant is required, then let me know a good place for that.

File: test/e2e/container_iface_name_test.go
Comment (vikas-goel) [reply test/e2e/container_iface_name_test.go]: I had to bring this back after deduplicating the tests by adding a loop. In the absence of "rm", the next iteration would conflict with the previous run and fail.

File: test/e2e/container_iface_name_test.go
Comment (Luap99) [inline test/e2e/container_iface_name_test.go:224]: These tests must work with netavark, since https://github.com/containers/podman/pull/21410 we will no longer have any CNI tests so this will never be run in CI.
What is the error reported?

File: test/e2e/container_iface_name_test.go
Comment (Luap99) [inline test/e2e/container_iface_name_test.go:91]: please remove this part from all your tests, they will clean up on their own and this juts bloats the lines with unnecessary code.

**Review by Luap99 — COMMENTED** [review]
Thanks, code looks good but we definitely need to have ipvlan tests working on netavark.

**Comment by vikas-goel** [comment]
> Thanks, code looks good but we definitely need to have ipvlan tests working on netavark.

It failed with a very clear statement that netavark does not support ipvlan.

File: test/e2e/container_iface_name_test.go
Comment (vikas-goel) [reply test/e2e/container_iface_name_test.go:91]: I had removed those statements but the test failed when it went in the next iteration that the container with the same name was already running. Definitely it didn’t auto cleanup at the end of the iteration. So, I had to add them back.

File: test/e2e/container_iface_name_test.go
Comment (vikas-goel) [reply test/e2e/container_iface_name_test.go:224]: It failed with a very clear statement that netavark does not support ipvlan.

**Comment by Luap99** [comment]
> It failed with a very clear statement that netavark does not support ipvlan.

What version are you using, the latest version support it and they should be in CI

**Comment by Luap99** [comment]
Lets see if https://github.com/containers/podman/pull/21530 works, if it does then it should work here too.

**Comment by vikas-goel** [comment]
> Lets see if #21530 works, if it does then it should work here too.

All ipvlan tests are failing in #21530. Will it make sense to get this PR in and enable ipvlan separately in #21530?

**Comment by vikas-goel** [comment]
After re-enabling ipvlan test for netavark, the same old issue happening.

```
# podman [options] run -d --network ipvlan5fd463d37d --name test quay.io/libpod/alpine:latest top
           Error: netavark: unknown network driver ipvlan
```

Any clues?

**Comment by vikas-goel** [comment]
System test failing randomly

sys remote fedora-38 root host boltdb
```
Failed tests (1):
 - 512 [500] podman run port forward range
```

**Comment by edsantiago** [comment]
Known flake, it's been hitting us very hard this past week (on f38 only). No point restarting it because the other failures are not flakes.

Looks like Debian has bad/old netavark (1.4).  We don't have `SkipIfDebian()`, but we do have `SkipIfRunc()`. Would you mind adding that to the failing tests, with a loud explanation?
```go
    SkipIfRunc("FIXME: Fails with netavark < 1.10. Reenable once Debian gets an update")
```

**Comment by vikas-goel** [comment]
How about I add `skipIfDebian()`? Will it be `info.Distribution == "debian"`?

**Review by Luap99 — APPROVED** [review]
/lgtm
/hold

**Pull request merged**
